### PR TITLE
Add Portkey LLMOps integration

### PR DESCRIPTION
**Pull request by vrushankportkey**
Integrating Portkey, which adds production features like caching, tracing, tagging, retries, etc. to langchain apps.

  - Dependencies: None
  - Twitter handle: https://twitter.com/portkeyai
  - test_portkey.py added for tests
  - example notebook added in new utilities folder in modules
  
 Also fixed a bug with OpenAIEmbeddings where headers weren't passing.

cc @baskaryan

### CODE REVIEW

File: langchain/embeddings/openai.py [hunk 1]
```diff
@@ -112,6 +113,7 @@ async def async_embed_with_retry(embeddings: OpenAIEmbeddings, **kwargs: Any) ->
 
     @_async_retry_decorator(embeddings)
     async def _async_embed_with_retry(**kwargs: Any) -> Any:
+        kwargs.update(embeddings._invocation_params)  # Add invocation params
```
[the remark this answers is not in the document]
Comment (vrushankportkey) [reply langchain/embeddings/openai.py:116]: Yes. For example, passing headers here: `response = OpenAIEmbeddings(headers={"x-portkey-api-key": "API_KEY"})` isn't working.

This is unique only to OpenAIEmbeddings on Python. For other models + on JS, it is working.

[the remark this answers is not in the document]
Comment (baskaryan) [reply langchain/embeddings/openai.py:116]: hm not able to recreate. if i mess with headers seems to propagate to top level commands (e.g. embed_query)

Screenshot 2023-07-18 at 5 18 47 PM

[the remark this answers is not in the document]
Comment (vrushankportkey) [reply langchain/embeddings/openai.py]: Ok, dropping that change for now - will verify from my end

**Comment by vrushankportkey** [comment]
@baskaryan fixed formartting erros. It should pass all checks now.

**Comment by baskaryan** [comment]
looks amazing, thanks @vrushankportkey!

**Pull request merged**